text/0143-versioning-and-stability.md

```diff
@@ -0,0 +1,130 @@
+
+**Never create a dependency conflict between packages which rely on different versions of OpenTelemetry. Avoid breaking all stable public APIs.**
+Backwards compatibility is a strict requirement. Instrumentation APIs cannot create a version conflict, ever. Otherwise, OpenTelemetry cannot be embedded in widely shared libraries, such as web frameworks. Code written against older versions of the API must work with all newer versions of the API. Transitive dependencies of the API cannot create a version conflict. The OpenTelemetry API cannot depend on "foo" if there is any chance that any library or application may require a different, incompatible version of "foo." A library using OpenTelemetry should never become incompaible with other libraries due to a version conflict in one of OpenTelemetry's dependencies.
+
+(Theoretically, APIs can be deprecated and eventually removed, but this is a process measured in years and we have no plans to do so.)
+
+**Allow for multiple levels of package stability within the same release.**
+Provide maintainers a clear process for developing new, experimental APIs alongside stable APIs. DIfferent packages within the same release may have different levels of stability. This means that an implementation wishing to release stable tracing today must ensure that experimental metrics are factored out in such a way that breaking changes to metrics API do not destabilize the trace API packages.
+
+## Relevant architecture
+![drawing](img/0143_cross_cutting.png)
+
+At the highest architectural level, OpenTelemetry is organized into signals. Each signal provides a specialized form of observability. For example, tracing, metrics, and baggage are three separate signals. Signals share a common subsystem – context propagation – but they function independently from each other.
+
+Each signal provides a mechanism for software to describe itself. A codebase, such as an API handler or a database client, takes a dependency on various signals in order to describe itself. OpenTelemetry instrumentation code is then mixed into the other code within that codebase. This makes OpenTelemetry a **cross-cutting concern** - a piece of software which must be mixed into many other pieces of software in order to provide value. Cross-cutting concerns, by their very nature, violate a core design principlie – seperation of concerns. As a result, OpenTelemetry requires extra care and attention to avoid creating issues for the codebases which depend upon these cross-cutting APIs.
+
+OpenTelemetry is designed to separate the portion of each signal which must be imported as cross-cutting concerns from the portions of OpenTelemetry which can be managed independently. OpenTelemetry is also designed to be a pluggable framework. To accomplish this these goals, each signal consists of four types of packages:
```

Comment:
> each signal consists of four types of packages

> every signal has its own, independent API package

For C++, we have single source package distribution consisting of api and sdk for all the signals - Thus a single version number for the release. Do we need to revisit it now ?

Reply:
No, please stick to a single version number if you can. Let me know how I can make it clearer.

There is a reasonable proposal that the SDK may make breaking changes faster than the API, and a separate version number for each would allow us to follow semver in a cleaner fashion. With one number, either the API looks more unstable than it is, or breaking changes to the SDK will be minor version bumps, making the SDK look more stable than it is. So please consider the versioning implications of following semver with a single version number, as we need input on this subject. I would love it if we could stick to one release, one version, but this is a legitimate concern.

Reply:
Just following up @lalitb. We've come to the conclusion that a single version number is untenable, as it would break semver in some fashion. We expect the SDK to update faster than the API, and have added different long term support grantees for the different components. 

So, please take this into account with how your are packaging your distribution, and future proof it for the event that you may need to make a major version bump to the SDK separately from the API. Also, the minor versions are no longer required to be in sync, so consider if it would be confusing at this point to bump the API minor version if a change was only made to the SDK. Also, please have a look at the support guarantees for the SDK and Contrib from an ABI compatibility standpoint.

Reply:
@tedsuo sure. Regarding versioning, we are discussing this internally, and would come up with a plan during our community meeting next week. We do understand that there are going to more frequent minor version dump in sdk, without any api level changes.
ABI stability challenges are different and tougher to detect in in C++ as compared to other languages. We do provide ABI stability at API level. But it is not provided by the interfaces the SDK provides as implementation hooks to vendor implementers, like exporters, processors, aggregators and propagators. This is something difficult to achieve without running some smart ABI stability detection tools, and not in our plan. The problem right now I see is that once we are bumped to V1.x, we need to increment the major version for any ABI breaking changes. We may need some flexibility here, but let's us discuss this more internally and see the better way to achieve it.